### PR TITLE
Consolidate package metadata

### PR DESCRIPTION
**Pull request by ericpre**
Similar as https://github.com/hyperspy/holospy/pull/14 and others

### Progress of the PR
- [x] Consolidate package metadata in `pyproject.toml`,
- [x] Tidy up outdated and unmaintained packaging files, 
- [x] Use `setuptools_scm` to set version in package at build time,
- [x] Simplify GitHub CI workflows 
- [x] Add python 3.12,
- [n/a] docstring updated (if appropriate),
- [x] update user guide (`releasing_guide.md`),
- [x] added tests (package and test workflow),
- [x] add changelog entry,
- [x] ready for review.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -27,7 +27,7 @@ jobs:
             LABEL: -oldest
           # test minimum requirement
           - os: ubuntu
-            PYTHON_VERSION: '3.8'
+            PYTHON_VERSION: '3.12'
             PIP_SELECTOR: '[tests, coverage]'
             LABEL: -minimum
```
Comment (jlaehne) [inline .github/workflows/tests.yml]: Duplicate section

Comment (ericpre) [reply .github/workflows/tests.yml]: Done!

File: pyproject.toml [hunk 2]
```diff
@@ -1,47 +1,203 @@
...
+    "Intended Audience :: Developers",
+    "Intended Audience :: Science/Research",
+    "License :: OSI Approved :: GNU General Public License v3 (GPLv3)",
+    "Operating System :: OS Independent",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
+    "Topic :: Scientific/Engineering",
```
Comment (jlaehne) [inline pyproject.toml:18]:
```suggestion
```

Comment (jlaehne) [reply pyproject.toml:18]: duplicate line

File: pyproject.toml [hunk 3]
```diff
@@ -1,47 +1,203 @@
...
+    "EMSA",
+    "FEI",
+    "ICA",
+    "PCA",
+    "PES",
+    "STEM",
+    "TEM",
+    "curve fitting",
+    "data analysis",
+    "dm3",
```
Comment (jlaehne) [inline pyproject.toml]:
```suggestion
```

File: pyproject.toml [hunk 4]
```diff
@@ -1,47 +1,203 @@
...
+    "electron energy loss spectroscopy",
+    "electron microscopy",
+    "emi",
```
Comment (jlaehne) [inline pyproject.toml]:
```suggestion
```

File: pyproject.toml [hunk 5]
```diff
@@ -1,47 +1,203 @@
...
+    "energy dispersive x-rays",
+    "hyperspectral",
+    "hyperspectrum",
+    "hyperspy",
+    "machine learning",
+    "microscopy",
+    "model",
+    "msa",
```
Comment (jlaehne) [inline pyproject.toml]:
```suggestion
```

File: pyproject.toml [hunk 6]
```diff
@@ -1,47 +1,203 @@
...
+    "multidimensional",
+    "numpy",
+    "python",
+    "quantification",
+    "scipy",
+    "ser",
```
Comment (jlaehne) [inline pyproject.toml]:
```suggestion
```

File: pyproject.toml [hunk 7]
```diff
@@ -1,47 +1,203 @@
...
+    "PES",
+    "STEM",
+    "TEM",
+    "curve fitting",
+    "data analysis",
+    "dm3",
+    "electron energy loss spectroscopy",
+    "electron microscopy",
+    "emi",
+    "energy dispersive x-rays",
```
Comment (jlaehne) [inline pyproject.toml]:
```suggestion
    "energy dispersive x-ray spectroscopy",
```

File: pyproject.toml (hunk 3 above)
Comment (jlaehne) [reply pyproject.toml]: Opened https://github.com/hyperspy/rosettasciio/pull/196

File: pyproject.toml [hunk 8]
```diff
@@ -1,47 +1,203 @@
...
+    "scikit-image>=0.18",
+    "scipy>=1.5.0",
+    "sympy>=1.6",
+    "tqdm>=4.9.0",
+    "traits>=4.5.0",
+]
+description = "Multidimensional data analysis toolbox"
+dynamic = ["version"]
+keywords = [
+    "EDX",
```
Comment (jlaehne) [inline pyproject.toml:50]: The keywords contain a selective set of the methods that can be addressed with hyperspy and its extensions. I would not throw them out, but add at least a selection of other methods:
- 4DSTEM
- EBSD / electron backscatter diffraction
- CL / cathodoluminescence
- PL / photoluminescence
- off-axis electron holography

File: pyproject.toml [hunk 9]
```diff
@@ -1,47 +1,203 @@
...
+    "EELS",
+    "EFTEM",
+    "EMSA",
+    "FEI",
```
Comment (jlaehne) [inline pyproject.toml]:
```suggestion
```

File: releasing_guide.md [hunk 10]
```diff
@@ -29,3 +26,8 @@ Follow-up
 - Tidy up and close corresponding milestone
 - A PR to the conda-forge feedstock will be created by the conda-forge bot
 
+Additional information
+======================
+
+The version is defined by ``setuptools_scm`` at build time when packaging HyperSpy. In editable
```
Comment (jlaehne) [inline releasing_guide.md]:
```suggestion
The version is defined by ``setuptools_scm`` at build time when packaging HyperSpy. In an editable
```

File: upcoming_changes/3268.maintenance.rst [hunk 11]
```diff
@@ -0,0 +1,5 @@
+Consolidate package metadata:
+- use ``pyproject.toml`` only
+- clean up unmaintained packaging files
+- use ``setuptools_scm`` to define version
+- add python 3.12 to test matrix
```
Comment (jlaehne) [inline upcoming_changes/3268.maintenance.rst]:
```suggestion
Consolidate package metadata:

- use ``pyproject.toml`` only
- clean up unmaintained packaging files
- use ``setuptools_scm`` to define version
- add python 3.12 to test matrix

```

Comment (jlaehne) [reply upcoming_changes/3268.maintenance.rst]: blank lines needed to fix list formatting

File: pyproject.toml [hunk 12]
```diff
@@ -1,47 +1,203 @@
...
+    "License :: OSI Approved :: GNU General Public License v3 (GPLv3)",
+    "Operating System :: OS Independent",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
+    "Topic :: Scientific/Engineering",
+    "Topic :: Scientific/Engineering",
+    "Topic :: Scientific/Engineering :: Physics",
```
Comment (jlaehne) [inline pyproject.toml:19]: Add the Physics topics also to extension libraries?
(included in https://github.com/hyperspy/rosettasciio/pull/196 )

**Review by jlaehne — APPROVED** [review]

File: pyproject.toml [hunk 13]
```diff
@@ -1,47 +1,203 @@
...
+]
+dev = [
+    "black",
+    "hyperspy[all]",
+    "hyperspy[coverage]",
+    "hyperspy[doc]",
+    "hyperspy[tests]",
+]
+doc = [
+    "pydata_sphinx_theme",
```
Comment (jlaehne) [inline pyproject.toml:110]: Already add ``"numpydoc"`` for compatibility with #3273?

Comment (ericpre) [reply pyproject.toml:110]: They will merge conflict anyway, so I prefer to leave it to #3273.

File: pyproject.toml [hunk 14]
```diff
@@ -1,47 +1,206 @@
...
-  showcontent = true
+package = "hyperspy"
+type = [
+    { directory = "api", name = "API changes", showcontent = true },
+    { directory = "bugfix", name = "Bug Fixes", showcontent = true },
+    { directory = "deprecation", name = "Deprecations", showcontent = true },
+    { directory = "doc", name = "Improved Documentation", showcontent = true },
+    { directory = "enhancements", name = "Enhancements", showcontent = true },
+    { directory = "maintenance", name = "Maintenance", showcontent = true },
+    { directory = "new", name = "New features", showcontent = true },
```
Comment (jlaehne) [inline pyproject.toml]:
```suggestion
    { directory = "new", name = "New features", showcontent = true },
    { directory = "enhancements", name = "Enhancements", showcontent = true },
    { directory = "bugfix", name = "Bug Fixes", showcontent = true },
    { directory = "api", name = "API changes", showcontent = true },
    { directory = "deprecation", name = "Deprecations", showcontent = true },
    { directory = "doc", name = "Improved Documentation", showcontent = true },
    { directory = "maintenance", name = "Maintenance", showcontent = true },
```

Comment (jlaehne) [reply pyproject.toml]: as proposed in https://github.com/hyperspy/rosettasciio/pull/197#issuecomment-1852678328

Comment (jlaehne) [reply pyproject.toml]: only wondering now, whether enhancements should go higher up in the list

Comment (jlaehne) [reply pyproject.toml]: now with enhancements on 2nd position

File: pyproject.toml [hunk 15]
```diff
@@ -1,47 +1,206 @@
...
-  [[tool.towncrier.type]]
-  directory = "api"
-  name = "API changes"
-  showcontent = true
-
-  [[tool.towncrier.type]]
-  directory = "maintenance"
-  name = "Maintenance"
-  showcontent = true
+package = "hyperspy"
```
Comment (jlaehne) [inline pyproject.toml:197]: In the other packages, this line is `package_dir`

Also the line `title_format = "{version} ({project_date})"` is missing?

Comment (ericpre) [reply pyproject.toml:197]:
```suggestion
package = "hyperspy"
title_format = "{version} ({project_date})"
```

For consistency, should we remove all the `v` (for example in `v2.0rc0`) from the release notes?

> In the other packages, this line is `package_dir`

According to https://towncrier.readthedocs.io/en/stable/configuration.html#extra-top-level-keys-for-python-projects, we should use `package`.

Comment (jlaehne) [reply pyproject.toml:197]: Well it seems to be two different things, but the extensions currently define only `package_dir`, while here only `package` is defined.

Comment (jlaehne) [reply pyproject.toml:197]: > For consistency, should we remove all the `v` (for example in `v2.0rc0`) from the release notes?

I always like consistency :-)

**Pull request merged**